### PR TITLE
Whole app rework

### PR DESCRIPTION
**Pull request by MarcinMajka**
Changes made:
- deleted unnecessary files
- fixed bugs related to clicking words and translations
- created the database of kanji, reading and glossary
- created a div for viewing the glossary, after a correct match has been made
- created a necessary info/link to JMdict, because of the license
- added timer

### CODE REVIEW

File: .DS_Store
Comment (aboutroots) [inline .DS_Store:1]: `.DS_Store` files are automatically added by MacOS . To keep the repo clean, add
```
.DS_Store
```
to the file called `.gitignore` and git will ignore them (we don't want this kind of files in the repo). https://www.atlassian.com/pl/git/tutorials/saving-changes/gitignore

But these files were already added - you need to remove them from git manually: https://stackoverflow.com/questions/1274057/how-do-i-make-git-forget-about-a-file-that-was-tracked-but-is-now-in-gitignore

File: JMdictFiles/.DS_Store
Comment (aboutroots) [inline JMdictFiles/.DS_Store:1]: same as above

File: JMdictFiles/JMdict_to_JSON.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import xml.etree.ElementTree as ET
+import json
+import re
+
+# Regular expression pattern to match kanji characters
+# [\u4e00-\u9faf] is a range for common kanji characters
+kanji_pattern = re.compile(r'[\u4e00-\u9faf]')
+
+# Parse the XML file
+tree = ET.parse('JMdict_e')
```
Comment (aboutroots) [inline JMdictFiles/JMdict_to_JSON.py]: do we have this file in the repo? There seems to be some inconsistency, lets think what is required for a fellow contributor to work:
1. If we have this `JMdict_to_JSON.py` script then we should have `JMdict_e` file to work with. One might argue that having the `JMdict.json` is now unnecessary because we can generate it on demand from the former files.
2. Alternatively, if we already have `JMdict.json`, why is there a need for `JMdict_to_JSON.py` when we cannot use it without `JMdict_e` ?

File: JMdictFiles/JMdict_to_JSON.py
Comment (aboutroots) [inline JMdictFiles/JMdict_to_JSON.py:1]: nice script :)

File: JMdictFiles/JMdict_to_JSON.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import xml.etree.ElementTree as ET
```
Comment (aboutroots) [inline JMdictFiles/JMdict_to_JSON.py]: Please add module docstring in the first line of the file, explaining the need and usage of this script.

File: JMdictFiles/parsing_JMDict.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import xml.etree.ElementTree as ET
```
Comment (aboutroots) [inline JMdictFiles/parsing_JMDict.py]: Please add module docstring in the first line of the file, explaining the need and usage of this script.

File: JMdictFiles/sameKanjiCounter.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+const { JM } = require('./dic');
```
Comment (aboutroots) [inline JMdictFiles/sameKanjiCounter.js]: again looks like we are missing some files in the repo? Note that empty folders will not be picked up by git.

File: JMdictFiles/sameKanjiCounter.js
Comment (aboutroots) [inline JMdictFiles/sameKanjiCounter.js:1]: Nice split into functions. Please add docstrings to module and functions to explain the usage and reasoning behind them.

PS these are good candidates for tests :thinking:

File: JMdictFiles/sameKanjiCounter.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        kanjiToCount[dict[i].kanji] = (kanjiToCount[dict[i].kanji] || 0) + 1;
+    }
+    return kanjiToCount;
+};
+
+const kanjiToCountWithAllKanjis = createKanjiToCountObject(JM);
+
+const createNonDistinctKanjiToCountObject = () => {
+    const nonUniqueKanjiToCount = {};
+    for (const kanji in kanjiToCountWithAllKanjis) {
```
Comment (aboutroots) [inline JMdictFiles/sameKanjiCounter.js]: instead of using a variable defined outside of the scope of the function, please pass the required data as arguments to the function.

Functions should be black boxes: 
```
---input---->[function]---output-->
```
if the function depends on stuff defined outside, you will have problems testing it in isolation - it is also easy to lost track of code executions and what is the actual state of the application (when your code grows).

If you find it convenient to have some local state shared among a couple functions, write a class instead.

File: JMdictFiles/sameKanjiCounter.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
+const { JM } = require('./dic');
+
+const createKanjiToCountObject = (dict) => {
+    const kanjiToCount = {};
+    for (let i = 0; i < dict.length; i++) {
+        kanjiToCount[dict[i].kanji] = (kanjiToCount[dict[i].kanji] || 0) + 1;
+    }
```
Comment (aboutroots) [inline JMdictFiles/sameKanjiCounter.js]: This works, although a more common practice among JS/TS developers is to make extensive use of `.forEach` , `.map` and `.reduce` functions to achieve the same result. Just make sure you can do both to understand the code that someone else might write.

File: dic.js
Comment (aboutroots) [inline dic.js:1]: The repo is growing in files - could you please add a `README.md` file? You could explain how to use the repo for fellow contributors. https://www.freecodecamp.org/news/how-to-write-a-good-readme-file/
and https://gist.github.com/DomPizzie/7a5ff55ffa9081f2de27c315f5018afc

File: script.js [hunk 7]
```diff
@@ -16,28 +16,38 @@ const shuffleArray = (array) => {
...
+let timerInterval;
 // -------------------
 
+const kanjiHiraganaGlossary = [];
+for(let i = 0; i < totalWordsInSessionCount; i++) {
+  const kanji = JM[i].kanji;
+  const hiragana = JM[i]['hiragana/katakana'];
+  const glossary = JM[i].glossary;
+  kanjiHiraganaGlossary.push([kanji, hiragana, glossary]);
+}
```
Comment (aboutroots) [inline script.js]: lets create a function for this logic - one should avoid logic in the global scope. Keep in mind that you will have to write tests for this app, and the easiest thing for testing is a plain function. The tests will look somehow like this:
```javascript
import { createKanjiHiraganaGlossary } from './functions.js';

test('should create glossary when input data is correct', () => {
  // Arrange
  const JM = { 
     // ...
  };
  const totalWords = 10;
  // Act
  const result = createKanjiHiraganaGlossary(JM, totalWords);
  // Assert
  expect(result).toMatchObject({ 
     // ...
  });
});
```
whereas if you leave logic in the global scope... good luck with testing :sweat_smile:

File: script.js [hunk 8]
```diff
@@ -52,22 +62,22 @@ const setupRound = (wordPairs, pairRenderLimitIndex) => {
 
   // create the divs for the words
   while (lastUsedPairIndex < pairRenderLimitIndex) {
-    const word = document.createElement("div");
+    const kanji = document.createElement("div");
     const wordValue = wordPairs[lastUsedPairIndex][0];
 
-    word.classList.add(`box`);
-    word.innerHTML = wordValue;
+    kanji.classList.add(`box`);
```
Comment (aboutroots) [inline script.js]: I would suggest using names like `left` and `right` or `column1` , `column2` or maybe `key` and `value` etc instead of kanji/hiragana. This way if you decide to add the ability to play in reverse (hiragana -> kanji) you don't have to rewrite all the code :D This goes also for function names etc.

This is formally known as decoupling of the business language (or domain language, such as "kanji") from UI language (such as "column")

We want to have the code as generic as possible, especially if this is a low cost (just a matter of decision). Useful framing to have in mind: "if I want to sell my code as a product, it should be able to suit multiple businesses" - maybe someone would like to use your code to learn French? etc etc

File: script.js [hunk 9]
```diff
@@ -97,91 +107,155 @@ function highlightElements(elements, className) {
...
 
 window.addEventListener("load", () => {
-  // We are starting the game when the page is loaded - before that, we don't have the divs to work with (they are not rendered yet)
-  // Create the initial state of the game - generate the divs with words and kanjis in HTML.
-  setupRound(wordPairs, pairCount);
-  // TODO: add a function to start the timer here.
+  // We are starting the game when the page is loaded - before that, we don't have the divs to work with (they are not rendered yet).
+  // Create the initial state of the game - generate the divs with kanji and hiragana/katakana in HTML.
+  setupRound(shuffledKanjiHiraganaGlossaries, pairsToRenderCount);
+  starTimer();
```
Comment (aboutroots) [inline script.js]: works nicely. Maybe next step would be to add a "start again" button for the whole game?

File: style.css [hunk 10]
```diff
@@ -1,19 +1,42 @@
+* {
+  margin: 0;
+  padding: 0;
+}
+
```
Comment (aboutroots) [inline style.css]: I see what you want to achieve here. Take a look at https://necolas.github.io/normalize.css/ which is a standard way of doing this :)

File: style.css [hunk 11]
```diff
@@ -45,3 +67,33 @@ body {
   border-color: #ff6b6b;
   color: #ff6b6b;
 }
+
+/* removes the possibility of reselecting the divs during animations */
```
Comment (aboutroots) [inline style.css]: I don't like this solution (although it is smart) - we should stick to using JS for logic, and ignore the pointer events in the listener function when appropriate conditions occur. You moved part of the logic to the styles, which make it very hard to test and not obvious for fellow developers (who will look at the JS in search for program logic)

**Review by aboutroots — CHANGES_REQUESTED** [review]
Great effort. Works nicely and looks great. Take a look at my comments below.

File: JMdictFiles/JMdict_to_JSON.py (hunk 1 above)
Comment (MarcinMajka) [reply JMdictFiles/JMdict_to_JSON.py]: I was on edge about deleting the files altogether, because I've put considerable time for them to work the way I wanted. If there wasn't a 50MB cap on files in github, `JMdict_e` would have been in the repo too.
I see 3 possible solutions:
1. Add a link to `JMdict_e` download page above line 10
2. Try to make `JMdict_e` weigh < 50MB
3. Untrack all files related to `JMdict_e` - least favourite of mine, because I made dic.js from a part of `JMdict_e`, which has like 130k+ records

Which one should we go with? My personal favourite is 1

Comment (MarcinMajka) [reply JMdictFiles/JMdict_to_JSON.py]: Decided to remove everything related to JMdict

File: .gitignore [hunk 12]
```diff
@@ -1 +1,4 @@
-
+.gitignore
+package-lock.json
```
Comment (aboutroots) [inline .gitignore]: do not put package-lock in the .gitignore - you want to share this with your coworkers/bots

File: README.md
Comment (aboutroots) [inline README.md:1]: :heart:

File: dic.js
Comment (aboutroots) [inline dic.js:1]: we will probably want to have this converted to a plain JSON file. This way we can treat it as a "static" file (omitted by bundler, e.g. webpack), because there is no need to process it.

File: functions.test.js [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+  };
+  test('["A", "A", "A", "B", "B", "B"]', () => {
+    const input = ["A", "A", "A", "B", "B", "B"];
+    const shuffledArray = shuffleArray([...input]);
+    // Expect shuffled array to not have the same order of elements
+    expect(shuffledArray).not.toEqual(input);
+    // Expect the count of each element be the same in both input and shuffledArray
+    expect(elementsToOccurrences(shuffledArray)).toEqual(
+      elementsToOccurrences(input)
+    );
```
Comment (aboutroots) [inline functions.test.js]: You created a function `elementsToOccurrences` - which is nice - but you introduced a new place for bugs (within this function!). 
I am exaggerating the problem here because `elementsToOccurrences` is quite simple, but you know what I mean - we have function A used to check function B, but how do we know that function A works? You should write tests for `elementsToOccurrences` too.

File: functions.test.js [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+  });
+  test('["A"]', () => {
+    const input = ["A"];
+    const shuffledArray = shuffleArray([...input]);
+    // Expect shuffled array to have the same order of elements (1 element)
+    expect(shuffledArray).toEqual(input);
+    // Expect the count of each element be the same in both input and shuffledArray
+    expect(elementsToOccurrences(shuffledArray)).toEqual(
+      elementsToOccurrences(input)
+    );
```
Comment (aboutroots) [inline functions.test.js]: This check is not needed, because it will always be true when the previous check is true.

File: functions.test.js [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+  });
+  test("[]", () => {
+    const input = [];
+    const shuffledArray = shuffleArray([...input]);
+    // Expect shuffled array to have the same order of elements (1 element)
+    expect(shuffledArray).toEqual(input);
+    // Expect the count of each element be the same in both input and shuffledArray
+    expect(elementsToOccurrences(shuffledArray)).toEqual(
+      elementsToOccurrences(input)
+    );
```
Comment (aboutroots) [inline functions.test.js]: same here. I like that you tested this edge case :muscle:

File: functions.test.js [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+// const JM = require('./dic');
+
+describe("Test shuffleArray", () => {
+  const elementsToOccurrences = (arr) => {
+    arr.reduce((obj, item) => {
+      obj[item] = (obj[item] || 0) + 1;
+      return obj;
+    }, {});
+  };
+  test('["A", "A", "A", "B", "B", "B"]', () => {
```
Comment (aboutroots) [inline functions.test.js:15]: Your tests could have more descriptive names. For example "Should shuffle an already sorted array"

File: functions.test.js
Comment (aboutroots) [inline functions.test.js:1]: very nice :) The general rule for this kind of unit test is to divide each test into 3 sections, like you did in your first examples, and the well known names for these sections are:
```
Arrange
Act
Assert
```
or 
```
Given
When
Then
```
The second ones are more useful for writing docstrings. Sticking to "triple A" (arrange-act-assert) tests is a good rule of thumb, because sometimes we are tempted to write too convoluted tests, check too many things etc.

In your example:
```javascript
/**
 * GIVEN an already sorted array
 * WHEN this array is provided as an argument to the function
 * THEN the resulting array should have different ordering 
 *    AND the number of occurrences of each item should stay the same
 */
test('["A", "A", "A", "B", "B", "B"]', () => {
    // Arrange
    const input = ["A", "A", "A", "B", "B", "B"];
    // Act
    const shuffledArray = shuffleArray([...input]);
    // Assert
    expect(shuffledArray).not.toEqual(input);
    expect(elementsToOccurrences(shuffledArray)).toEqual(
      elementsToOccurrences(input)
);
```

File: functions.test.js [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    ]);
+  });
+});
+
+describe("Test formatTime", () => {
+  test("More than 9", () => {
+    expect(formatTime(10)).toBe(10);
+    expect(formatTime(31)).toBe(31);
+    expect(formatTime(54)).toBe(54);
+  });
```
Comment (aboutroots) [inline functions.test.js]: There is a special syntax in `jest` that streamlines writing this type of simple tests. It makes use of `.each` method:
```javascript
test.each([10, 31, 54])("More than 9: %i", (value) => {
  expect(formatTime(value)).toBe(value);
});

// OR more explicit format using backticks ``

test.each`
  value
  ${10}
  ${31}
  ${54}
`("More than 9: $value", ({ value }) => {
  expect(formatTime(value)).toBe(value);
});
```
in the first variation you can put whatever JS objects you want (dicts too) which can be handy. In the second "table" syntax you can have many columns, which is often more readable. Here is an example from another project:
```javascript
 test.each`
    code      | index | expected
    ${''}     | ${1}  | ${''}
    ${'abcd'} | ${0}  | ${'a'}
    ${'abcd'} | ${1}  | ${'b'}
    ${'abcd'} | ${2}  | ${'c'}
    ${'abcd'} | ${3}  | ${'d'}
    ${'abcd'} | ${4}  | ${''}
  `('inputValue returns correct value', ({ code, index, expected }) => {
    const wrapper = shallowMountComponent();
    wrapper.setData({ code });
   
    const result = wrapper.vm.inputValue(index);
    
    expect(result).toBe(expected);
  });
  ```

File: matchMadness.html [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+    </div>
+    <div class="container">
+      <div id="glossaryContainer">
+        <p id="leftValueRightValue"></p>
+        <p id="glossary"></p>
+      </div>
+    </div>
+    <div class="buttonAndLicenseContainer">
+      <div class="buttonContainer">
+        <a href="matchMadness.html" class="button restartGame">R E S T A R T</a>
```
Comment (aboutroots) [inline matchMadness.html]: spacing in words is better done with CSS https://developer.mozilla.org/en-US/docs/Web/CSS/letter-spacing?retiredLocale=pl

File: script.js [hunk 19]
```diff
@@ -94,168 +122,214 @@ const setupRound = (wordPairs, pairRenderLimitIndex) => {
...
-    for (let i = lastUsedPairIndex - pairsToRenderCount; i < lastUsedPairIndex; i++) {
-      console.log("i: ", i)
-      console.log("hiragana: " + hiragana + " shuffledKanjiHiraganaGlossaries[i][1]: " + shuffledKanjiHiraganaGlossaries[i][1])
-      if (kanji === shuffledKanjiHiraganaGlossaries[i][0] && hiragana === shuffledKanjiHiraganaGlossaries[i][1]) {
-        expectedHiragana = shuffledKanjiHiraganaGlossaries[i][1];
+    for (
+      let i = lastUsedTripletIndex - pairsToRenderCount;
+      i < lastUsedTripletIndex;
+      i++
+    ) {
```
Comment (aboutroots) [inline script.js:205]: Make sure to format automatically with prettier. You can fine-tune prettier by adding a configuration file to the root of the repository (read more in docs https://prettier.io/docs/en/configuration.html)

File: script.js [hunk 20]
```diff
@@ -94,168 +122,214 @@ const setupRound = (wordPairs, pairRenderLimitIndex) => {
  * @param className - the class to add and then remove
  */
 function highlightElements(elements, className) {
-  elements.forEach((element) => element.classList.add(className));
+  // Highlighted elements should be disabled during the animation
+  elements.forEach((element) => {
+    element.classList.add(className);
+    element.style.pointerEvents = "none";
```
Comment (aboutroots) [inline script.js]: this is not really what I had in mind :D You still have the logic in CSS, you just dynamically add the CSS here. Its better, but not ideal.

What I had in mind was to add `if` condition in the function that reacts to the events (In your case, this function is `checkIfMatch`. Something like

```javascript
const checkIfMatch = (event) => {
  const clickedElement = event.target; 
  // block any action for elements that are already highlighted
  const elementIsBlocked = // ...add your logic - check the classes etc
  if (elementIsBlocked) return;
  /// ...

**Review by aboutroots — APPROVED** [review]
Good work! I'm happy to see nice tests, even if we are not covering everything yet. You have my approval as there are no big issues, just make sure to read all the comments and apply wherever necessary. :v:

File: functions.test.js [hunk 21]
```diff
@@ -1,114 +1,112 @@
...
-});
+    { kanji: '火', 'hiragana/katakana': 'ひ', glossary: 'fire' },
+    { kanji: '水', 'hiragana/katakana': 'みず', glossary: 'water' },
+    { kanji: '木', 'hiragana/katakana': 'き', glossary: 'wood' },
+    { kanji: '土', 'hiragana/katakana': 'ど', glossary: 'earth' },
+    { kanji: '心臓', 'hiragana/katakana': 'しんぞう', glossary: 'heart' },
+    { kanji: '空', 'hiragana/katakana': 'そら', glossary: 'sky' },
+  ]
+  test('3 expected entries', () => {
+    console.log('test 3')
```
Comment (aboutroots) [inline functions.test.js]: remove console logs

**Review by aboutroots — APPROVED** [review]

**Review by aboutroots — APPROVED** [review]
Schitts Creek Perfect
